index.d.ts

```diff
@@ -41,15 +43,18 @@ export interface Action {
  *
  * *Do not put API calls into reducers.*
  *
- * @template S State object type.
+ * @template S The type of state consumed and produced by this reducer.
+ * @template A The type of actions the reducer can potentially respond to.
  */
-export type Reducer<S> = <A extends Action>(state: S | undefined, action: A) => S;
+export type Reducer<S = {}, A extends Action = Action> = (state: S | undefined, action: A) => S;
```

Comment:
Still, do we really want to have `S` default to `{}`? Why?

Reply:
I was trying to keep this PR as backwards-compatible as possible, but if that's not important I'm more than happy to remove it.

Reply:
Sorry, I don't get how it aids backward compatibility. Previous `Reducer` type did have `S` as a required parameter.

Reply:
Hm, I appear to have been confused when I made this comment. I was thinking of the state parameter to the `Dispatch` type, which is completely unnecessary but I was keeping around for backwards compatibility.

The argument for defaulting `S` here is the same as for `ReducersMapObject` and `Store` below... so that we can declare type constraints like `R extends Reducer` when we don't care about the type arguments.

Reply:
Thanks, this makes sense. The only thing that bothers me: if the state type of reducer includes `null`, that reducer won't be assignable to `Reducer` (without parameters).

A very simplified example:
```ts
type Reducer<S = {}> = (state: S | undefined) => S;

const r: Reducer = (state: null | undefined = null): null => {
  return state;
}
```
Gives
```
Type '(state?: null | undefined) => null' is not assignable to type 'Reducer<{}>'.
  Types of parameters 'state' and 'state' are incompatible.
    Type '{} | undefined' is not assignable to type 'null | undefined'.
      Type '{}' is not assignable to type 'null | undefined'.
        Type '{}' is not assignable to type 'null'.
```

Works fine after setting `S = {} | null`.

Reply:
Yeah, I'm coming to think the default type should probably be either  be `S = any` or `S = {} | null | undefined`... a "top" type, since `S` is in a contravariant position.

Reply:
I think you're right, except that `undefined` for the state is prohibited by Redux. Yet I'm not sure if we gain any type safety by excluding it.

As far as I know, `{} | null | undefined` is the same as `any`.

Reply:
Remember though, `undefined` is prohibited to return from a reducer but not as argument, which is passed with the initial call.

Reply:
It's not quite the same, because for example if

```ts
declare x: {} | null | undefined
x.foo // this is a type error
declare y: any
y.foo // this is fine
```

`any` isn't really a type, it's more like a directive not to perform type checking. I'd say `{} | null` is the most "correct" thing to use here, but a lot of people tend to use `any` in situations like this because it's easy on the eyes. I could go either way.

Reply:
> Remember though, `undefined` is prohibited to return from a reducer but not as argument, which is passed with the initial call.

The definition covers this since the state argument is of type `S | undefined`.

Reply:
> It's not quite the same

You are right, I used the wrong word. They are equivalent in terms of assignability.